### PR TITLE
Initial attempt at Windows support

### PR DESCRIPTION
**Pull request by RyanMelenaNoesis**
Uses [NuGet librdkafka - redistributable](https://www.nuget.org/packages/librdkafka.redist/) rather than compiling librdkafka for Windows.  Should address #45.

Also replaced `usleep` function with cross-platform `std::this_thread::sleep_for` alternative.

### CODE REVIEW

File: src/workers.cc [hunk 1]
```diff
@@ -378,14 +378,14 @@ void KafkaConsumerConsumeLoop::Execute(const ExecutionMessageBus& bus) {
       // when in consume loop mode
       // Randomise the wait time to avoid contention on different
       // slow topics
-      usleep(static_cast<int>(rand_r(&m_rand_seed) * 1000 * 1000 / RAND_MAX));
+      std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(rand() * 1000 / RAND_MAX)));
```
Comment (vsimonian) [inline src/workers.cc]: GNU `rand()` is not thread-safe, but `rand_r(*seedp)` is. Only MSVC's `rand()` is thread-safe. Therefore it seems that `rand()` should only be used in this context if it is being compiled by MSVC, not on all platforms.

Comment (RyanMelenaNoesis) [inline src/workers.cc]: Good catch @vsimonian.  C++ isn't my strong suit so thanks for the tip, I'll check in a change with an IFDEF momentarily.

**Comment by mattness** [comment]
Thanks for the PR and your interest in the project, folks. We always love receiving contributions from the community!

Please remember that our first priority is ensure the stability and maintainability of the Linux build.  With that in mind, I have a couple objections to this PR as it stands right now:

1. `std:this_thread` is a C++11 construct.  `node-rdkafka` currently doesn't rely on C++11, and I'd prefer not to increase the compiler requirement if it can be avoided.
1. The addition of the `download` and `decompress` dependencies have no use in the library other than to support a platform-specific build. For applications that use node-rdkafka and don't run on Windows, these dependencies add additional attack surface area (as well as litter to the directory tree) with **zero** benefit to the application.
1. The `nuget-dep.js` and `node-gyp rebuild` steps in package.json are unnecessary for non-Windows installs, yet will still run and produce more litter. The reliance on nuget.org also introduces a requirement for internet connectivity that did not previously exist, and is commonly not readily available in enterprise environments or clean-room build environments.
1. I can't see anything that ensures the version of the binary being downloaded from nuget matches the revision of the source in the submodule. There's no way for a user of `node-rdkafka` to ensure that the downloaded binary doesn't contain malicious code, or for testers/maintainers to know that it is even built from the intended revision of librdkafka.

In general, I'm against downloading a pre-built binary (there's a reason node-gyp exists, after all). While it may be the path of least resistance to a "Works on my Machine" build, it's also a path of great burden for those who will be providing user support and maintaining the codebase.

I'd be more willing to accept a PR for Windows support where librdkafka was built from the submodule source at install time, just as the Linux build works now.

**Comment by RyanMelenaNoesis** [comment]
With regards to your concerns:

1. I was working off the recommendation of @webmakersteve [here](https://github.com/Blizzard/node-rdkafka/issues/45#issuecomment-321307346) to avoid the need for a polyfill.  I would be open to surrounding the new imports and `std:this_thread` uses with IFDEFs or returning to the polyfill approach used by @vsimonian as well.
2. If you were open to it, I would be happy to convert the current nuget download/extract process into a bash script that could be executed in the npm scripts prepack step.  Alternatively, you could simply include the windows binaries as part of the repository.
3. Either of the alternatives I offered in point 2 would eliminate the need for internet connectivity for end-users.
4. There is nothing currently that controls the version of the binary being downloaded from nuget but it would be trivial to specify the nuget package version to download or use in the alternatives provided in point 2.  With regards to ensuring the downloaded binary doesn't contain malicious code, I think the wide-spread usage of NuGet, node-pre-gyp, pre-minified JS, CDNs, package managers, etc points to the concern being significantly overblown.

I'm happy to work further to address some of the concerns mentioned but requiring librdkafka to be built form source is frankly a bridge too far for me.  Given time constraints, if that is the bar that must be cleared I'd prefer to simply maintain my existing fork and point our package.json to it.

Personally, I'd have to think that non-ideal Windows support has to be better than no Windows support and I have a hard time seeing how the changes (even as implemented today) represent a 
> great burden for those who will be providing user support and maintaining the codebase

but obviously the final choice is not mine.

**Comment by mattness** [comment]
Thanks so much for your effort thus far. We really do appreciate it. However, I need to insist that librdkafka is built from the submodule source.

I'll go ahead and close this PR for now, given that

> requiring librdkafka to be built from source is frankly a bridge too far

If time and inclination permits in the future, please feel free to reopen it with the needed changes.

**Pull request closed without merging**